go-controller/pkg/util/external_gw_conntrack.go

```diff
@@ -0,0 +1,315 @@
+		return false
+	}
+
+	for _, iface := range ifaces {
+		if isDown(iface) {
+			continue
+		}
+		if !hasAddressInNetwork(iface) {
+			continue
+		}
+		return &iface, nil
+	}
+	return nil, errors.New("no usable interface found")
+}
+
+func readNDPMsg(msg ndp.Message) (mac net.HardwareAddr, err error) {
+	// Expect a neighbor advertisement message with a target link-layer
+	// address option.
+	na, ok := msg.(*ndp.NeighborAdvertisement)
+	if !ok {
+		err = fmt.Errorf("message is not a neighbor advertisement: %T", msg)
+		return
+	}
+	if len(na.Options) != 1 {
+		err = fmt.Errorf("expected one option in neighbor advertisement")
+		return
+	}
+	lla, ok := na.Options[0].(*ndp.LinkLayerAddress)
+	if !ok {
+		err = fmt.Errorf("option is not a link-layer address: %T", msg)
+		return
+	}
+	return lla.Addr, nil
+}
+
+// getIPv6MacOnIface tries to resolve as many ips as possible.
+// Errors that prevent only 1 ip from being resolved are logged and not returned.
+// When an error is returned, some MACs may still be resolved and returned too.
+func getIPv6MacOnIface(info *ifaceWithTargetIPs) ([]net.HardwareAddr, error) {
+	// Set up an *ndp.Conn, bound to this interface's link-local IPv6 address.
+	c, _, err := ndp.Listen(info.iface, ndp.LinkLocal)
+	if err != nil {
+		return nil, fmt.Errorf("failed to dial NDP connection: %v", err)
+	}
+	defer c.Close()
+	for _, resolveIP := range info.ips {
+		target, err := netip.ParseAddr(resolveIP)
+		if err != nil {
+			klog.Errorf("Failed to ParseAddr %v: %v", resolveIP, err)
+			continue
+		}
+		// Use target's solicited-node multicast address to request that the target
+		// respond with a neighbor advertisement.
+		snm, err := ndp.SolicitedNodeMulticast(target)
+		if err != nil {
+			klog.Errorf("Failed to determine solicited-node multicast address: %v", err)
+			continue
+		}
+
+		// Build a neighbor solicitation message, indicate the target's link-local
+		// address, and also specify our source link-layer address.
+		m := &ndp.NeighborSolicitation{
+			TargetAddress: target,
+			Options: []ndp.Option{
+				&ndp.LinkLayerAddress{
+					Direction: ndp.Source,
+					Addr:      info.iface.HardwareAddr,
+				},
+			},
+		}
+
+		// Send the multicast message and wait for a response.
+		if err = c.WriteTo(m, nil, snm); err != nil {
+			klog.Errorf("Failed to write neighbor solicitation: %v", err)
+			continue
+		}
+	}
+	ipsToFind := sets.New[string](info.ips...)
+	macs := []net.HardwareAddr{}
+
+	// set max attempts to msgTries to limit the number of messages waiting for expected reply
+	for i := 0; i < msgTries*len(info.ips); i++ {
+		if err = c.SetReadDeadline(time.Now().Add(msgTimeout)); err != nil {
+			return macs, fmt.Errorf("failed to set read deadline: %w", err)
+		}
+		msg, _, ip, err := c.ReadFrom()
+		if err != nil {
+			// if some ips are resolved and others are not available anymore, return macs
+			// when no more messages are received
+			return macs, fmt.Errorf("failed to read NDP message: %v", err)
```

Comment:
so if we don't get a message in 500ms, then we would timeout and return error. We also have msgTries=10, but that doesnt have anything to do with the time we wait for a message, since a single message timeout signals failure. So I guess the msgTries is just if we are getting messages, we allow up to 10*the number of messages we sent to be the wrong messages.

This seems a little over complicated. I might just set an overall timeout that scales by how many ips we are looking for, and run the loop reading packets until that expires or we find all our MACs.

Reply:
I just think structuring your loop with an overall timeout is more obvious and cleaner, like:

```
trozet@fedora:~/go/src/github.com/ovn-org/ovn-kubernetes$  git diff
diff --git a/go-controller/pkg/util/external_gw_conntrack.go b/go-controller/pkg/util/external_gw_conntrack.go
index 541d8f7d0..9e16fa8c2 100644
--- a/go-controller/pkg/util/external_gw_conntrack.go
+++ b/go-controller/pkg/util/external_gw_conntrack.go
@@ -133,7 +133,8 @@ func getIPv6MacOnIface(info *ifaceWithTargetIPs) ([]net.HardwareAddr, error) {
        macs := []net.HardwareAddr{}
 
        // set max attempts to msgTries to limit the number of messages waiting for expected reply
-       for i := 0; i < msgTries*len(info.ips); i++ {
+       maxDuration := time.Duration(msgTries*len(info.ips)) * msgTimeout
+       for start := time.Now(); time.Since(start) < maxDuration; {
                if err = c.SetReadDeadline(time.Now().Add(msgTimeout)); err != nil {
                        return macs, fmt.Errorf("failed to set read deadline: %w", err)
                }
```

Reply:
ah I see, so you suggest to use msgTimeout as a multiplier, but it is not exactly the same as current code.
If there are no messages on the interface, we will timeout in 500 ms, but we want to also ensure that if there are a lot of messages, but the one we are waiting for doesn't come, we exit at some point. When there is lots of messages, we will receive them fairly quickly (based on my testing ^ ~ 5 ms), therefore to timeout for 100 ips, it will take us ~ 5ms*10(msgTries)*100 ips = 5000 ms = 5 sec.
If we multiply on a msg timeout, it will take 100 ips * 500ms (message timeout) = 50 sec, so I was concerned that this is too much, but it is definitely safer. If we are ok with this longer timeout, I will happily change it

Reply:
Right, that was my intention here. If there is some network outage for like 500ms, we would timeout after not receiving any messages, then potentially miss responses. I'd rather be on the safer side right now and take too long, and we can tweak it later if there are issues.

Reply:
I have just realized that in your suggestion you have `msgTries` multiplier, which I removed (otherwise it would make my example go to 500 sec), lmk if you actually want to leave it